Added set method for Keys, set method for Flags

GerRudi opened the pull request:
Added set method for Keys
Added new method "Flags" to set Flags for Accelerator, get method not yet included

Edit: btw, I suck at coding, it works but feel free to improve my code. :)

GerRudi commented:
get method for the flags is now included in ee581bf

dblock commented on Docs/Accelerator.md:
```diff
@@ -42,6 +42,18 @@ BEGIN
+| ------------- |:-------------:| -----:|
+| `VK_TAB`      | key top press for Accelerator | Virtual Key or ASCII character |
+| `41008`      | command / identifier to execute      |   Integer-Value |
+| `VIRTKEY , NOINVERT , CONTROL` |   options    |    string, entries separated with ',' |
```
No need to align column 3 to the right.

Cleanup the Capitals vs. lowercase things. In the last column for example you say `string`, which should probably be `String`. 

`Integer-Value` doesn't sound like a type to me, maybe just `Integer`? 

I think it's either `Virtual Key or ASCII Character` or `virtual key or ASCII character`

dblock commented on Docs/Accelerator.md:
```diff
@@ -42,6 +42,18 @@ BEGIN
+| `VK_TAB`      | key top press for Accelerator | Virtual Key or ASCII character |
+| `41008`      | command / identifier to execute      |   Integer-Value |
+| `VIRTKEY , NOINVERT , CONTROL` |   options    |    string, entries separated with ',' |
+For more Information, please look at the [MSDN reference](https://msdn.microsoft.com/en-us/library/windows/desktop/aa380610(v=vs.85).aspx "ACCELERATOR resources").
```
Change `please look at` to `see`. Information should be lowercase.

dblock commented on Docs/Accelerator.md:
```diff
@@ -57,7 +69,45 @@ Console.WriteLine("AcceleratorResource: {0}, {1}", rc.Name, rc.TypeName);
 
+Creating a new Accelerator
+----------------------------
+To add a new `Accelerator` to an `AcceleratorResource`, you need to create the `Accelerator` first and assign a key, command and flags.
```
you can remove "you need"

dblock commented on Docs/Accelerator.md:
````diff
@@ -57,7 +69,45 @@ Console.WriteLine("AcceleratorResource: {0}, {1}", rc.Name, rc.TypeName);
+To add a new `Accelerator` to an `AcceleratorResource`, you need to create the `Accelerator` first and assign a key, command and flags.
+The Example below creates a new `Accelerator` that executes command **1337** when **NUMPAD2** is pressed.
+``` csharp
+  Accelerator acNew = new Accelerator();
````
Line up `acNew....` to the left.

dblock commented on Docs/Accelerator.md:
```diff
@@ -57,7 +69,45 @@ Console.WriteLine("AcceleratorResource: {0}, {1}", rc.Name, rc.TypeName);
 
 In order to add an accelerator into an executable (`.exe` or `.dll`) create a new instance of `AcceleratorResource` or load an existing one and add instances of `Accelerator` to its `Accelerators` collection. 
+
+The following example loads the accelerators from explorer.exe, removes accelerator 6, and adds a new accelarator instead
```
Needs a period at the end.

dblock commented on Docs/Accelerator.md:
```diff
@@ -57,7 +69,45 @@ Console.WriteLine("AcceleratorResource: {0}, {1}", rc.Name, rc.TypeName);
+    foreach (AcceleratorResource rc in ri[Kernel32.ResourceTypes.RT_ACCELERATOR])
+    {
+        Console.WriteLine(rc);
+        rc.Accelerators.RemoveAt(5);//zero-based index - removing Acceleator 6
```
Add some spaces around `//`.

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -45,7 +45,58 @@ public string Key
+            {
+                try//will work if given string is a Virtual Key e.g. "VK_NUMPAD1"
+                {
+                    uint i = (uint)Enum.Parse(typeof(User32.VirtualKeys), value);
```
Use `TryParse`, avoid using the exception to handle the error case.

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -45,7 +45,58 @@ public string Key
+            get
+            {
+                uint flags = _accel.fVirt;
+                List<string> sA=new List<string>();
```
Space around `=`.

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -45,7 +45,58 @@ public string Key
+        /// When using set method, several flags have to be split with ',' e.g. "VIRTKEY, NOINVERT, CONTROL"
+        /// get method will return flags in the same way, e.g. "VIRTKEY,SHIFT,ALT"
+        /// </summary>
+        public string Flags
```
You probably also want a method that returns and sets the value of `fVirt` directly?

dblock replied:
I don't think this should be a string method that parses comma-separated values. It should be an unsigned integer property that you can assign `AcceleratorVirtualKey` values to. Or maybe it should be a collection of these. Either way, String seems unnecessarily complicated and very error prone here.

dblock commented:
Thanks. I made a bunch of comments.

Most importantly this needs tests and an entry in CHANGELOG, please.

dblock commented:
@GerRudi lmk when this is ready to re-review, maybe squash commits too.

GerRudi commented:
Thanks for your feedback!

- Removed try/catch from Accelerator.Key (instead using Enum.IsDefined)
- Accelerator.Flags now working with User32.AcceleratorVirtualKey instead of strings
- Fixed formattings & typos + corrected documentation for new Accelerator.Flags
- Added AcceleratorTests.cs to ResourceLibUnitTests
- Added addFlag & removeFlag methods to remove single flags
- Added this pull request to changelog

I'm done with changes for now, please feel free to re-review and I'd be happy to receive feedback again.

If you can tell me how to squash commits via GitHub.com, I can merge them.
(I'm currently not using git...)

dblock commented on CHANGELOG.md:
```diff
@@ -3,6 +3,9 @@
 **New Features**
 
 * [#40](https://github.com/dblock/resourcelib/pull/40): Added `Resource.Save(string filename, IEnumerable<Resource> resources)` for saving multiple resources in one batch - [@thoemmi](https://github.com/thoemmi).
+* [#46](https://github.com/dblock/resourcelib/pull/46): Added set-method for `Accelerator.Keys`, 
```
What's up with new lines and plusses ;) 

Make this three entires with the same pull request number and your name.

Added set method for for `Accelerator.Keys`
Added `Accelerator.Flags`
Added `Accelerator.addFlag(User32.AcceleratorVirtualKey)` and `Accelerator.removeFlag(User32.AcceleratorVirtualKey)`

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -45,12 +45,45 @@ public string Key
+        }
+
+        /// <summary>
+        /// AcceleratorVirtualKey representation of the Accelerator bit-flags.
```
I think most of this comment is no longer true.

GerRudi replied:
Hm, maybe I'm not seeing it or I wrote it wrong but for me the comment seems correct.
By the way I'm not sure if this isn't too detailed for a summary? 
I have no experience in that stuff...

dblock replied:
You're right. Not sure what I was thinking :)

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -65,6 +98,26 @@ public UInt32 Command
+        /// <param name="f">the User32.AcceleratorVirtualKey flag to add</param>
+        public void addFlag(User32.AcceleratorVirtualKey f)
+        {
+            if (!((_accel.fVirt & (uint)f) == (uint)f))
```
You can set this via a `|=`, see http://stackoverflow.com/questions/93744/most-common-c-sharp-bitwise-operations-on-enums for various options of how to implement that.

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -65,6 +98,26 @@ public UInt32 Command
+        /// <param name="f">the User32.AcceleratorVirtualKey flag to remove</param>
+        public void removeFlag(User32.AcceleratorVirtualKey f)
+        {
+            if ((_accel.fVirt & (uint)f) == (uint)f)
```
Same here, you shouldn't need an if.

dblock commented on Source/ResourceLibUnitTests/AcceleratorTests.cs:
```diff
@@ -0,0 +1,65 @@
+using System.IO;
+using Vestris.ResourceLib;
+
+namespace Res_sample
```
This isn't a test, it's a sample. Run NUnit tests on this project, and extend https://github.com/dblock/resourcelib/blob/master/Source/ResourceLibUnitTests/AcceleratorResourceTests.cs with additional tests for various simple scenarios that you're implementing. There're lots of examples there.

dblock commented:
Cool. See above. Don't worry about squashing.

GerRudi commented:
I made some changes for you to review again.
- Accelerator.addFlag & Accelerator.removeFlag now using bitwise operators
- Fixed formatting for Changelog
- deleted AcceleratorTest.cs, instead added NUnit test for new methods in AcceleratorResourceTests.cs
- NUnit test passed with 0 errors

Thanks!

dblock commented on Source/ResourceLib/Accelerator.cs:
```diff
@@ -65,6 +98,24 @@ public UInt32 Command
+        /// <param name="f">the User32.AcceleratorVirtualKey flag to add</param>
+        public void addFlag(User32.AcceleratorVirtualKey f)
+        {
+            _accel.fVirt |= (UInt16)f;
```
Much better!

dblock commented on Source/ResourceLibUnitTests/AcceleratorResourceTests.cs:
```diff
@@ -32,5 +32,65 @@ public void TestLoadAcceleratorResources()
+                    }
+                    catch (Exception e)
+                    {
+                        Console.WriteLine("Error: " + e.ToString());
```
For a unit test, you want to "assert" things, not have a pass-through fail like this. Look at the `TestLoadAcceleratorResources` above for an example. Maybe read up a bit on NUnit.

I think you want to break this up in multiple tests.

* Adding one accelerator.
* Modifying one accelerator.
* Removing one accelerator.
* Doing something on multiple accelerators.

Think of `TestLoadAcceleratorResources` as a start. Now that there's a lot more functionality in accelerators supported write the smallest tests possible for each of those functions. Together they cover the entire functionality of the code that you're adding, across multiple tests.

dblock commented:
Getting very close! Lets get those tests to be really great and it will be mergable.

dblock commented:
Bump @GerRudi, want to finish this?

thoemmi commented:
@GerRudi Please use PascalCasing for `AddFlag` and `RemoveFlag`, as recommended in Microsoft's [Naming Guidelines](https://msdn.microsoft.com/en-us/library/ms229043(v=vs.110).aspx).
